File: typescript/src/data-structures/linked-list/linked-list.ts
export class LinkedListNode<T = any> {
  data: T;
  next: LinkedListNode<T> = null;
  constructor(data: T) {
    this.data = data;
  }
}

export class LinkedList<T = any> {
  head: LinkedListNode<T>;
  tail: LinkedListNode<T>;
  length: number;
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  public push(val: T): LinkedListNode<T> {
    const node = new LinkedListNode(val);
    if (!this.head) {
      this.head = node;
    } else {
      this.tail.next = node;
    }

    this.tail = node;
    this.length++;
    return node;
  }

  public pop(): LinkedListNode<T> {
    if (!this.head) return undefined;

    let currentNode = this.head;
    let newTailNode = this.tail;

    while (currentNode.next) {
      newTailNode = currentNode;
      currentNode = currentNode.next;
    }

    this.tail = newTailNode;
    this.tail.next = null;
    this.length--;

    if (this.length === 0) {
      // last element popped, reset list
      this.head = null;
    }
    return currentNode;
  }

  public shift(): LinkedListNode<T> {
    if (!this.head) return undefined;

    const returnedHead = this.head;

    this.head = this.head.next;

    this.length--;
    if (this.length === 0) {
      // last element shifted, reset list
      this.tail = null;
      this.head = null;
    }
    return returnedHead;
  }

  public unshift(val: T): LinkedListNode<T> {
    const node = new LinkedListNode(val);

    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      const oldHead = this.head;
      this.head = node;
      this.head.next = oldHead;
    }

    this.length++;

    return this.head;
  }

  public get(position: number): LinkedListNode<T> {
    let currentNode = this.head;
    let i = 0;
    if (position < 0 || position >= this.length) return null;
    while (i !== position) {
      currentNode = currentNode.next;
      i++;
    }

    return currentNode;
  }

  public set(position: number, val: T): boolean {
    const node = this.get(position);
    if (!node) return false;
    node.data = val;
    return true;
  }

  public insert(position: number, val: T) {
    if (position === 0) return !!this.unshift(val);
    else if (position === this.length - 1) return !!this.push(val);
    else {
      const prevNode = this.get(position - 1);
      if (!prevNode) return false;
      const nextNode = prevNode.next;
      const newNode = new LinkedListNode(val);
      prevNode.next = newNode;
      newNode.next = nextNode;
    }

    this.length++;
    return true;
  }

  public remove(position: number) {
    if (position === 0) return this.shift();
    else if (position === this.length - 1) return this.pop();
    else {
      const prevNode = this.get(position - 1);
      if (!prevNode) return false;
      const thisNode = this.get(position);
      prevNode.next = thisNode.next;
      this.length--;
      return thisNode;
    }
  }

  public reverse() {
    let node = this.head;
    this.head = this.tail;
    this.tail = node;

    let next;
    let prev = null;

    while (node !== null) {
      next = node.next;
      node.next = prev;
      prev = node;
      node = next;
    }

    return this;
  }

  public sort() {
    let currentNode = this.head;
    let index = null;
    let temp: T;

    if (this.head === null) {
      return;
    } else {
      while (currentNode !== null) {
        index = currentNode.next;
        while (index !== null) {
          if (currentNode.data > index.data) {
            temp = currentNode.data;
            currentNode.data = index.data;
            index.data = temp;
          }
          index = index.next;
        }
        currentNode = currentNode.next;
      }
    }
  }

  public display() {
    const result = [];

    let node = this.head;

    while (node) {
      result.push(node.data);
      node = node.next;
    }

    return result;
  }
}

File: go/linked-list/linked-list.go
package linkedlist

import "errors"

type LinkedListNode struct {
	Data int
	Next *LinkedListNode
}

type LinkedList struct {
	Head   *LinkedListNode
	Tail   *LinkedListNode
	Length int
}

func (list *LinkedList) Push(element int) LinkedListNode {
	node := LinkedListNode{Data: element}

	if list.Head == nil {
		list.Head = &node
	} else {
		list.Tail.Next = &node
	}

	list.Tail = &node
	list.Length++
	return node
}

func (list *LinkedList) Pop() (*LinkedListNode, error) {
	if list.Head == nil {
		return nil, errors.New("linked list is empty")
	}

	currentNode := list.Head
	newTailNode := list.Tail

	for currentNode.Next != nil {
		newTailNode = currentNode
		currentNode = currentNode.Next
	}

	list.Tail = newTailNode
	list.Tail.Next = nil
	list.Length--

	if list.Length == 0 {
		list.Head = nil
	}

	return currentNode, nil
}

func (list *LinkedList) Shift() (*LinkedListNode, error) {
	if list.Head == nil {
		return nil, errors.New("linked list is empty")
	}

	returnedHead := list.Head

	list.Head = list.Head.Next

	list.Length--

	if list.Length == 0 {
		list.Tail = nil
		list.Head = nil
	}

	return returnedHead, nil
}

func (list *LinkedList) Unshift(element int) (*LinkedListNode, error) {
	node := LinkedListNode{Data: element}

	if list.Head == nil {
		list.Head = &node
		list.Tail = &node
	} else {
		oldHead := list.Head
		list.Head = &node
		list.Head.Next = oldHead
	}

	list.Length++

	return list.Head, nil
}

func (list *LinkedList) Get(position int) *LinkedListNode {
	currentNode, i := list.Head, 0

	if position < 0 || position >= list.Length {
		return nil
	}

	for i != position {
		currentNode = currentNode.Next
		i++
	}

	return currentNode
}

func (list *LinkedList) Set(position int, element int) bool {
	node := list.Get(position)

	if node == nil {
		return false
	} else {
		node.Data = element
		return true
	}
}

func (list *LinkedList) Insert(position int, element int) bool {
	if position == 0 {
		_, err := list.Unshift(element)
		return err == nil
	} else if position == list.Length-1 {
		list.Push(element)
		return true
	} else {
		prevNode := list.Get(position - 1)
		if prevNode == nil {
			return false
		}
		nextNode, newNode := prevNode.Next, LinkedListNode{Data: element}

		prevNode.Next = &newNode
		newNode.Next = nextNode
	}

	list.Length++
	return true
}

func (list *LinkedList) Remove(position int) *LinkedListNode {
	if position == 0 {
		node, err := list.Shift()
		if err != nil {
			return node
		} else {
			return nil
		}
	} else if position == list.Length-1 {
		node, err := list.Pop()
		if err != nil {
			return node
		} else {
			return nil
		}
	} else {
		prevNode := list.Get(position - 1)
		if prevNode == nil {
			return nil
		}
		thisNode := list.Get(position)
		prevNode.Next = thisNode.Next
		list.Length--
		return thisNode
	}
}

func (list *LinkedList) Reverse() *LinkedList {
	node := list.Head
	list.Head = list.Tail
	list.Tail = node

	var next *LinkedListNode
	var prev *LinkedListNode = nil

	for node != nil {
		next = node.Next
		node.Next = prev
		prev = node
		node = next
	}

	return list
}

func (list *LinkedList) Sort() {
	currentNode := list.Head
	var index *LinkedListNode
	var temp int

	if list.Head == nil {
		return
	} else {
		for currentNode != nil {
			index = currentNode.Next
			for index != nil {
				if currentNode.Data > index.Data {
					temp = currentNode.Data
					currentNode.Data = index.Data
					index.Data = temp
				}
				index = index.Next
			}
			currentNode = currentNode.Next
		}
	}
}

func (list *LinkedList) Display() []int {
	var result []int
	node := list.Head

	for node != nil {
		result = append(result, node.Data)
		node = node.Next
	}

	return result
}
